Added objects and functions for SKDrawable

charlenni opened the pull request:
This is code for the https://github.com/mono/SkiaSharp/issues/726 (Missing SkDrawable) issue

Changed samples for testing. Should be reverted.

Not sure, if the memory management is correct. Sorry. Not very firm in C++.

mikekinsman commented:
OPS Build status updates of commit _[e1c0178](https://github.com/charlenni/SkiaSharp/pull/739/commits/e1c0178a62b5ccfe3bea42e490bc1b5c429c0753)_: 

:clock10: **Preparing**: average preparing time is _1 min(s) 52 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[e1c0178](https://github.com/charlenni/SkiaSharp/pull/739/commits/e1c0178a62b5ccfe3bea42e490bc1b5c429c0753)_: 

:clock10: **Incremental building**: average incremental building time is _22 sec(s)_

charlenni commented:
Sorry, not clear, why the commit 2397595 is inside this PR. Shouldn't be there. How could I remove this?

mikekinsman commented:
OPS Build status updates of commit _[e1c0178](https://github.com/charlenni/SkiaSharp/pull/739/commits/e1c0178a62b5ccfe3bea42e490bc1b5c429c0753)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[.gitmodules](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/.gitmodules) | :white_check_mark:Succeeded | |
[binding/Binding/SKAbstractDrawable.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKAbstractDrawable.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKCanvas.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKCanvas.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKDrawable.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKDrawable.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SkiaApi.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SkiaApi.cs) | :white_check_mark:Succeeded | |
[externals/skia](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/externals/skia) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Class1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Class1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Form1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Form1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj) | :white_check_mark:Succeeded | |
[source/SkiaSharp.Views.Forms/SkiaSharp.Views.Forms.Shared/SKTouchEventArgs.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/source/SkiaSharp.Views.Forms/SkiaSharp.Views.Forms.Shared/SKTouchEventArgs.cs) | :white_check_mark:Succeeded | |
[source/SkiaSharp.Views.Forms/SkiaSharp.Views.Forms.UWP/SKTouchHandler.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/source/SkiaSharp.Views.Forms/SkiaSharp.Views.Forms.UWP/SKTouchHandler.cs) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C12%5C20%5C81ca5486-8c50-0a94-05c4-486e23e99ecf%5CPullRequest%5C201812201334439517-739%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=OBGKzE6ksJBGEuWx%2BMX%2FSvjbhvm9QhkMuWhWiFD6rmc%3D&st=2018-12-20T13%3A31%3A51Z&se=2019-01-20T13%3A36%3A51Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

mattleibow commented:
The master branch was a bit behind the development branch (I was testing there). I just updated master, so you should be able to update your branch for all the goodness.

mikekinsman commented:
OPS Build status updates of commit _[955c7e7](https://github.com/charlenni/SkiaSharp/pull/739/commits/955c7e735f1e9c2659ac9d29da3c6105e299434f)_: 

:clock10: **Preparing**: average preparing time is _1 min(s) 52 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[955c7e7](https://github.com/charlenni/SkiaSharp/pull/739/commits/955c7e735f1e9c2659ac9d29da3c6105e299434f)_: 

### :x: Validation status: errors


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[]() | :x:Error | | [Details](#user-content-)

<a name=""></a>
### []()
 - **[Error]** Cannot sync git repo to specified state. Please make sure the pull request is mergeable. Details: Git merge results in conflict, repository remote url: https://github.com/mono/SkiaSharp, working directory: W:\shny-s, target branch: master, target branch commit: 826f6a25ecba556c593ce11c272dc990c0d73dbf, source branch: a29d8f504285ca21560ae9f6e2ebb4ef/SkDrawable, source branch commit: .

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C12%5C20%5C81ca5486-8c50-0a94-05c4-486e23e99ecf%5CPullRequest%5C201812201407201875-739%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=BI7l6siFcnowoPP7qX1nnizS6KMGzHbR8nvTjIwFhDo%3D&st=2018-12-20T14%3A02%3A38Z&se=2019-01-20T14%3A07%3A38Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace SkiaSharp
+{
+	public abstract class SKDrawable : SKAbstractDrawable
```
Because there is a `SKDrawable` from the native code, we should rather name this `SKManagedDrawable` to both match the C++ types, but also so we can keep the inheritance the same.

We can add a `SKDrawable` that is also abstract (so that we cannot new it up) and an internal constructor (so that we can prevent users from creating any direct types) - they need to use the `SKManagedDrawable`.

mikekinsman commented:
OPS Build status updates of commit _[00b8017](https://github.com/charlenni/SkiaSharp/pull/739/commits/00b80177551103bfbf11a9c63bba5e0862e018de)_: 

:clock10: **Preparing**: average preparing time is _1 min(s) 52 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[00b8017](https://github.com/charlenni/SkiaSharp/pull/739/commits/00b80177551103bfbf11a9c63bba5e0862e018de)_: 

:clock10: **Incremental building**: average incremental building time is _22 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[00b8017](https://github.com/charlenni/SkiaSharp/pull/739/commits/00b80177551103bfbf11a9c63bba5e0862e018de)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[.gitmodules](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/.gitmodules) | :white_check_mark:Succeeded | |
[binding/Binding/SKAbstractDrawable.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKAbstractDrawable.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKCanvas.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKCanvas.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKDrawable.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKDrawable.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SkiaApi.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SkiaApi.cs) | :white_check_mark:Succeeded | |
[externals/skia](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/externals/skia) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Class1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Class1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Form1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Form1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C12%5C20%5C81ca5486-8c50-0a94-05c4-486e23e99ecf%5CPullRequest%5C201812201414499594-739%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=eZZ9QvVpOforou7e%2FdVDZivgADfmaMIqhX%2Bot8fC4VA%3D&st=2018-12-20T14%3A11%3A53Z&se=2019-01-20T14%3A16%3A53Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

mattleibow commented on binding/Binding/SKAbstractDrawable.cs:
```diff
@@ -0,0 +1,125 @@
+
+		// delegate declarations
+		[UnmanagedFunctionPointer (CallingConvention.Cdecl)]
+		internal delegate IntPtr draw_delegate (IntPtr managedDrawablePtr, IntPtr canvasPtr);
```
```suggestion
		internal delegate void draw_delegate (IntPtr managedDrawablePtr, IntPtr canvasPtr);
```

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,64 @@
+			base.Dispose (disposing);
+		}
+
+		public uint GenerationID => SkiaApi.sk_manageddrawable_get_generation_id (Handle);
```
```suggestion
		public uint GenerationId => SkiaApi.sk_manageddrawable_get_generation_id (Handle);
```

.NET naming

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,64 @@
+			SkiaApi.sk_manageddrawable_draw (Handle, canvas.Handle, ref matrix);
+		}
+
+		public SKPicture NewPictureSnapshot ()
```
I am not sure what the best naming would be:

 - `ToPicture` (this is a nice one)
 - `AsPicture` (may imply that there is a simple cast, and not a totally disconnect)
 - `CreatePicture` (may be confusing with the past use of static constructor types)
 - `Snapshot` (used by `SKSurface` to indicate capturing the current state in an immutable form, so we have consistency here)

Any other ideas?

mattleibow replied:
I think we should go with `ToPicture` here, unless you have any other ideas.

```suggestion
		public SKPicture ToPicture ()
```

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,64 @@
+			SkiaApi.sk_manageddrawable_notify_drawing_changed (Handle);
+		}
+
+		protected override SKPicture OnNewPictureSnapshot ()
```
😍 I really like what you did here!

mattleibow commented on samples/Basic/Desktop/SkiaSharpSample/Class1.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace SkiaSharpSample
+{
+	public class Class1 : SKDrawable
```
This might be better served in the tests. If you need to check what was drawn, you can do something like I do and just check a few pixels' color.

charlenni replied:
That's correct. I onlky use this for fast test, if all could be compiled. I could remove this later.

mattleibow commented on samples/Basic/Desktop/SkiaSharpSample/Form1.cs:
```diff
@@ -38,6 +38,12 @@ private void skiaView_PaintSurface(object sender, SKPaintSurfaceEventArgs e)
+			canvas.DrawDrawable(drawable, 0, 0);
+			drawable.NewPictureSnapshot();
+			var id = drawable.GenerationID;
+			drawable.NotifyDrawingChanged();
```
This might be something for the tests.

mikekinsman commented:
OPS Build status updates of commit _[3abc2a5](https://github.com/charlenni/SkiaSharp/pull/739/commits/3abc2a59edcc37f6e8bec69b117c93869156c82c)_: 

:clock10: **Preparing**: average preparing time is _1 min(s) 51 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[3abc2a5](https://github.com/charlenni/SkiaSharp/pull/739/commits/3abc2a59edcc37f6e8bec69b117c93869156c82c)_: 

:clock10: **Incremental building**: average incremental building time is _22 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[3abc2a5](https://github.com/charlenni/SkiaSharp/pull/739/commits/3abc2a59edcc37f6e8bec69b117c93869156c82c)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[.gitmodules](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/.gitmodules) | :white_check_mark:Succeeded | |
[binding/Binding/SKCanvas.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKCanvas.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKDrawable.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKDrawable.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKPictureRecorder.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKPictureRecorder.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SkiaApi.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SkiaApi.cs) | :white_check_mark:Succeeded | |
[externals/skia](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/externals/skia) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Class1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Class1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Form1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Form1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C12%5C20%5C81ca5486-8c50-0a94-05c4-486e23e99ecf%5CPullRequest%5C201812202035256917-739%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=WkjOlh3JkBRagyKqIilQXUMeH764Va2u9Eqt7u1%2F2PE%3D&st=2018-12-20T20%3A32%3A27Z&se=2019-01-20T20%3A37%3A27Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,164 @@
+		[UnmanagedFunctionPointer (CallingConvention.Cdecl)]
+		internal delegate void getBounds_delegate (IntPtr managedDrawablePtr, out SKRect rect);
+		[UnmanagedFunctionPointer (CallingConvention.Cdecl)]
+		internal delegate SKPicture newPictureSnapshot_delegate (IntPtr managedDrawablePtr);
```
I believe this has to be `IntPtr` / `sk_picture_t` because this is an interop function. The equivalent is to pass the handle around:

```suggestion
		internal delegate IntPtr newPictureSnapshot_delegate (IntPtr managedDrawablePtr);
```

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,164 @@
+		}
+
+		[MonoPInvokeCallback(typeof(newPictureSnapshot_delegate))]
+		private static SKPicture NewPictureSnapshotInternal (IntPtr managedDrawablePtr)
```
This has to be a blittable type - `IntPtr` for the handle:

```suggestion
		private static IntPtr NewPictureSnapshotInternal (IntPtr managedDrawablePtr)
```

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,164 @@
+		[MonoPInvokeCallback(typeof(newPictureSnapshot_delegate))]
+		private static SKPicture NewPictureSnapshotInternal (IntPtr managedDrawablePtr)
+		{
+			return GetObject<SKPicture> (AsManagedDrawable (managedDrawablePtr).OnNewPictureSnapshot ().Handle);
```
```suggestion
			return AsManagedDrawable (managedDrawablePtr).OnNewPictureSnapshot ().Handle;
```

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,164 @@
+			: base (SkiaApi.sk_manageddrawable_new (), owns)
+		{
+			if (Handle == IntPtr.Zero) {
+				throw new InvalidOperationException ("Unable to create a new SKAbstractDrawable instance.");
```
```suggestion
				throw new InvalidOperationException ("Unable to create a new SKDrawable instance.");
```

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,164 @@
+			if (AsManagedDrawable (ptr, out var target)) {
+				return target;
+			}
+			throw new ObjectDisposedException ("SKAbstractDrawable: " + ptr);
```
```suggestion
			throw new ObjectDisposedException ("SKDrawable: " + ptr);
```

mattleibow commented on binding/Binding/SKDrawable.cs:
```diff
@@ -0,0 +1,164 @@
+			base.Dispose (disposing);
+		}
+
+		public uint GenerationID => SkiaApi.sk_drawable_get_generation_id (Handle);
```
.NET naming styles:

```suggestion
		public uint GenerationId => SkiaApi.sk_drawable_get_generation_id (Handle);
```

charlenni commented:
One Problem we now have is, that SKDrawable isn‘t anymore abstract. You suggested to make the ctors internal. Would this help?

charlenni commented:
Ok, tested it. It seems, that it isn't possible to create a instance of ````SKDrawable````. So this is good.

Some points for renaming NewPictureSnapshot:
* You find this name in the Skia documentation
* You could write an own event OnNewPictureSnapshot() which is called when calling NewPictureSnapshot. So, if you call the one ToPicture() and the other OnNewPictureSnapshot() than it could get problematic. In this case, we should rename OnNewPictureSnapshot() too. Any good idea? OnGetPicture()?

mikekinsman commented:
OPS Build status updates of commit _[546add4](https://github.com/charlenni/SkiaSharp/pull/739/commits/546add4889d3a28fad888fb5cf6515c0b86b562c)_: 

:clock10: **Preparing**: average preparing time is _1 min(s) 50 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[546add4](https://github.com/charlenni/SkiaSharp/pull/739/commits/546add4889d3a28fad888fb5cf6515c0b86b562c)_: 

:clock10: **Incremental building**: average incremental building time is _22 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[546add4](https://github.com/charlenni/SkiaSharp/pull/739/commits/546add4889d3a28fad888fb5cf6515c0b86b562c)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[.gitmodules](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/.gitmodules) | :white_check_mark:Succeeded | |
[binding/Binding/SKCanvas.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKCanvas.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKDrawable.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKDrawable.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKPictureRecorder.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKPictureRecorder.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SkiaApi.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SkiaApi.cs) | :white_check_mark:Succeeded | |
[externals/skia](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/externals/skia) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Class1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Class1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/Form1.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/Form1.cs) | :white_check_mark:Succeeded | |
[samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/samples/Basic/Desktop/SkiaSharpSample/SkiaSharpSample.csproj) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C12%5C21%5C81ca5486-8c50-0a94-05c4-486e23e99ecf%5CPullRequest%5C201812212039155679-739%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=cD2hm1fHY5BAxHhDkaFR7JOWs1yf3Yn2mphhGpbLTzM%3D&st=2018-12-21T20%3A36%3A22Z&se=2019-01-21T20%3A41%3A22Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

mattleibow commented:
I see... Names, eh?
Well, we could also go the `Snapshot()` (like `SKSurface`) and then `OnSnapshot` route... ?

mattleibow commented:
I think `Snapshot` is cool. More descriptive - it takes a snapshot of the current drawing. Just like it does for the surface. And, we don't need to have "picture" in the name as there is only one type of snapshot that this can make. I just had a look at the C++ code, and Google is using `makeImageSnapshot` in there, and Miguel renamed it to just `Snapshot`:

```csharp
public SKImage Snapshot () =>
    GetObject<SKImage> (SkiaApi.sk_surface_new_image_snapshot (Handle));
```

Looks like Miguel did not like the name either :)

charlenni commented:
I now used Snapshot() and OnSnapshot(). That should be good enough. But the internal functions are still named after the original Skia function names.

How is the documantation made? I thought, it is made by comments in the source code. But this isn't the case. So, where should I add the description of SKDrawable?

mattleibow commented:
build

mattleibow commented:
Nice!
So yeah, we keep the native methods the same name to match the C++ API - because we are not the only consumers.

Docs, that is a bit of a process as I usually wait for CI and then run 2 commands - first download all the assets for all the platforms, then run mdoc to generate all the various API docs and diffs. But, Let me see if I can create a new cake task just to generate the minimum for adding summaries and that. CI appears to be having issues accessing the hard disks. I really, really need to move this over to Azure DevOps to get a clean bot each time.

charlenni commented:
This means, that I don‘t write any documentation?

mattleibow commented:
Yeah :)
I tried to get something to work, but I need to bump the versions. And, for some reason both mdoc and the comparer just stopped working :(

I need to fix the scripts a bit more... This is good though, I can now see all the failings of the process.

mattleibow commented:
@charlenni I see there is an issue with the compilation:
https://jenkins.mono-project.com/blue/organizations/jenkins/SkiaSharp-Pulls/detail/SkiaSharp-Pulls/130/pipeline

```
../../../src/c/sk_drawable.cpp: In function 'void sk_drawable_draw(sk_drawable_t*, sk_canvas_t*, const sk_matrix_t*)':
../../../src/c/sk_drawable.cpp:33:54: error: taking address of temporary [-fpermissive]
     AsDrawable(d)->draw(AsCanvas(c), &AsMatrix(matrix));
                                                      ^
```

You may have to copy the matrix to a local variable.

mikekinsman commented:
OPS Build status updates of commit _[98159bf](https://github.com/charlenni/SkiaSharp/pull/739/commits/98159bfa9736e544f4e491c5d098a46d72a2f70e)_: 

:clock10: **Preparing**: average preparing time is _1 min(s) 49 sec(s)_

charlenni commented:
@mattleibow That's very strange, because, if I compile Skia on my machine, I don't have any problems.

I now removed the ````const```` before matrix parameter. Hope, that this removes the error. Could not test it (see above).

Removed the test from sample project.

mikekinsman commented:
OPS Build status updates of commit _[98159bf](https://github.com/charlenni/SkiaSharp/pull/739/commits/98159bfa9736e544f4e491c5d098a46d72a2f70e)_: 

:clock10: **Incremental building**: average incremental building time is _21 sec(s)_

mikekinsman commented:
OPS Build status updates of commit _[98159bf](https://github.com/charlenni/SkiaSharp/pull/739/commits/98159bfa9736e544f4e491c5d098a46d72a2f70e)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[.gitmodules](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/.gitmodules) | :white_check_mark:Succeeded | |
[binding/Binding/SKCanvas.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKCanvas.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKDrawable.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKDrawable.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SKPictureRecorder.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SKPictureRecorder.cs) | :white_check_mark:Succeeded | |
[binding/Binding/SkiaApi.cs](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/binding/Binding/SkiaApi.cs) | :white_check_mark:Succeeded | |
[externals/skia](https://github.com/charlenni/SkiaSharp/blob/SkDrawable/externals/skia) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C12%5C26%5C81ca5486-8c50-0a94-05c4-486e23e99ecf%5CPullRequest%5C201812261453412601-739%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=kvllFBpTdZ1M6VoGEh4SAsbwNA3W78sZyFFtwey7qt8%3D&st=2018-12-26T14%3A50%3A57Z&se=2019-01-26T14%3A55%3A57Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.
